File: Corte_2/EjerciciosOrdenamientos/TiposOrdenamientoClase.py
import time
# En este codigo se van a escribir los distintos tipos de ordenamiento como funciones
#Tambien se va a agregar cuantos intercambios se realizan en cada ordenamiento, el tiempo que se demora y las comparaciones que hace

def bubble_sort(arr):
    n = len(arr)
    intercambios = 0
    comparaciones = 0
    for i in range(n):
        swapped = False
        for j in range(0, n-i-1):
            comparaciones += 1
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
                swapped = True
                intercambios += 1
        if not swapped:
            break
    return arr, intercambios, comparaciones

def selection_sort(arr):
    n = len(arr)
    intercambios = 0
    comparaciones = 0
    for i in range(n):
        min_idx = i
        for j in range(i+1, n):
            comparaciones += 1
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
            intercambios += 1
    return arr, intercambios, comparaciones

def insertion_sort(arr):
    n = len(arr)
    intercambios = 0
    comparaciones = 0
    for i in range(1, n):
        key = arr[i]
        j = i-1
        while j >= 0 and key < arr[j]:
            comparaciones += 1
            arr[j+1] = arr[j]
            j -= 1
            intercambios += 1
        if j >= 0:
            comparaciones += 1
        arr[j+1] = key
    return arr, intercambios, comparaciones

def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]

        L, left_intercambios, left_comparaciones = merge_sort(L)
        R, right_intercambios, right_comparaciones = merge_sort(R)

        intercambios = left_intercambios + right_intercambios
        comparaciones = left_comparaciones + right_comparaciones

        i = j = k = 0

        while i < len(L) and j < len(R):
            comparaciones += 1
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
            intercambios += 1

        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1
            intercambios += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1
            intercambios += 1

        return arr, intercambios, comparaciones
    else:
        return arr, 0, 0

def quick_sort(arr):
    intercambios = 0
    comparaciones = 0

    def _quick_sort(arr, low, high):
        nonlocal intercambios, comparaciones
        if low < high:
            pi, pi_intercambios, pi_comparaciones = partition(arr, low, high)
            intercambios += pi_intercambios
            comparaciones += pi_comparaciones
            _quick_sort(arr, low, pi - 1)
            _quick_sort(arr, pi + 1, high)

    def partition(arr, low, high):
        intercambios_local = 0
        comparaciones_local = 0
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            comparaciones_local += 1
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
                intercambios_local += 1
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        intercambios_local += 1
        return i + 1, intercambios_local, comparaciones_local

    _quick_sort(arr, 0, len(arr) - 1)
    return arr, intercambios, comparaciones

def heap_sort(arr):
    intercambios = 0
    comparaciones = 0

    def heapify(arr, n, i):
        nonlocal intercambios, comparaciones
        largest = i
        l = 2 * i + 1
        r = 2 * i + 2

        if l < n:
            comparaciones += 1
            if arr[l] > arr[largest]:
                largest = l

        if r < n:
            comparaciones += 1
            if arr[r] > arr[largest]:
                largest = r

        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            intercambios += 1
            heapify(arr, n, largest)

    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        intercambios += 1
        heapify(arr, i, 0)

    return arr, intercambios, comparaciones

def bucket_sort(arr):
    intercambios = 0
    comparaciones = 0

    if len(arr) == 0:
        return arr, intercambios, comparaciones

    bucket_count = len(arr)
    max_value = max(arr)
    min_value = min(arr)
    buckets = [[] for _ in range(bucket_count)]

    for num in arr:
        index = (num - min_value) * (bucket_count - 1) // (max_value - min_value)
        buckets[index].append(num)
        intercambios += 1

    sorted_arr = []
    for bucket in buckets:
        bucket.sort()
        comparaciones += len(bucket) * (len(bucket) - 1) // 2
        sorted_arr.extend(bucket)

    return sorted_arr, intercambios, comparaciones

datos = [64, 25, 12, 22, 11, 90, 45, 33]

start_time = time.time()
print("Bubble Sort:", bubble_sort(datos.copy()))
print("Numero de intercambios:", bubble_sort(datos.copy())[1])
print("Numero de comparaciones:", bubble_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

start_time = time.time()
print("Selection Sort:", selection_sort(datos.copy()))
print("Numero de intercambios:", selection_sort(datos.copy())[1])
print("Numero de comparaciones:", selection_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

start_time = time.time()
print("Insertion Sort:", insertion_sort(datos.copy()))
print("Numero de intercambios:", insertion_sort(datos.copy())[1])
print("Numero de comparaciones:", insertion_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

start_time = time.time()
print("Quick Sort:", quick_sort(datos.copy()))
print("Numero de intercambios:", quick_sort(datos.copy())[1])
print("Numero de comparaciones:", quick_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

start_time = time.time()
print("Heap Sort:", heap_sort(datos.copy()))
print("Numero de intercambios:", heap_sort(datos.copy())[1])
print("Numero de comparaciones:", heap_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

start_time = time.time()
print("Bucket Sort:", bucket_sort(datos.copy()))
print("Numero de intercambios:", bucket_sort(datos.copy())[1])
print("Numero de comparaciones:", bucket_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

start_time = time.time()
print("Merge Sort:", merge_sort(datos.copy()))
print("Numero de intercambios:", merge_sort(datos.copy())[1])
print("Numero de comparaciones:", merge_sort(datos.copy())[2])
end_time = time.time()
print("Tiempo transcurrido:", end_time - start_time)

File: Corte_2/EjerciciosOrdenamientos/TiposOrdenamientoClase.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

using namespace std;    

void bubble_sort(vector<int>& arr, int& intercambios, int& comparaciones) {
    int n = arr.size();
    intercambios = 0;
    comparaciones = 0;
    for (int i = 0; i < n; i++) {
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            comparaciones++;
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
                intercambios++;
            }
        }
        if (!swapped) {
            break;
        }
    }
}

void selection_sort(vector<int>& arr, int& intercambios, int& comparaciones) {
    int n = arr.size();
    intercambios = 0;
    comparaciones = 0;
    for (int i = 0; i < n; i++) {
        int min_idx = i;
        for (int j = i + 1; j < n; j++) {
            comparaciones++;
            if (arr[j] < arr[min_idx]) {
                min_idx = j;
            }
        }
        if (min_idx != i) {
            swap(arr[i], arr[min_idx]);
            intercambios++;
        }
    }
}

void insertion_sort(vector<int>& arr, int& intercambios, int& comparaciones) {
    int n = arr.size();
    intercambios = 0;
    comparaciones = 0;
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && key < arr[j]) {
            comparaciones++;
            arr[j + 1] = arr[j];
            j--;
            intercambios++;
        }
        if (j >= 0) {
            comparaciones++;
        }
        arr[j + 1] = key;
    }
}
void merge_sort(vector<int>& arr, int left, int right, int& intercambios, int& comparaciones) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        merge_sort(arr, left, mid, intercambios, comparaciones);
        merge_sort(arr, mid + 1, right, intercambios, comparaciones);

        int n1 = mid - left + 1;
        int n2 = right - mid;
        vector<int> L(n1);
        vector<int> R(n2);
        for (int i = 0; i < n1; i++)
            L[i] = arr[left + i];
        for (int j = 0; j < n2; j++)
            R[j] = arr[mid + 1 + j];

        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2) {
            comparaciones++;
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            intercambios++;
            k++;
        }

        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
            intercambios++;
        }

        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
            intercambios++;
        }
    }
}

void quick_sort(vector<int>& arr, int low, int high, int& intercambios, int& comparaciones) {
    if (low < high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            comparaciones++;
            if (arr[j] <= pivot) {
                i++;
                swap(arr[i], arr[j]);
                intercambios++;
            }
        }
        swap(arr[i + 1], arr[high]);
        intercambios++;
        int pi = i + 1;
        quick_sort(arr, low, pi - 1, intercambios, comparaciones);
        quick_sort(arr, pi + 1, high, intercambios, comparaciones);
    }
}

void heapify(vector<int>& arr, int n, int i, int& intercambios, int& comparaciones) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    if (largest != i) {
        swap(arr[i], arr[largest]);
        intercambios++;
        heapify(arr, n, largest, intercambios, comparaciones);
    }
}

void heap_sort(vector<int>& arr, int& intercambios, int& comparaciones) {
    int n = arr.size();
    intercambios = 0;
    comparaciones = 0;
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i, intercambios, comparaciones);
    }
    for (int i = n - 1; i > 0; i--) {
        swap(arr[i], arr[0]);
        intercambios++;
        heapify(arr, i, 0, intercambios, comparaciones);
    }
}

void bucket_sort(vector<int>& arr, int& intercambios, int& comparaciones) {
    int n = arr.size();
    if (n <= 0)
        return;

    int max_val = *max_element(arr.begin(), arr.end());
    int bucket_count = max_val / 10 + 1;
    vector<vector<int>> buckets(bucket_count);

    for (int i = 0; i < n; i++) {
        int idx = arr[i] / 10;
        buckets[idx].push_back(arr[i]);
        intercambios++;
    }

    arr.clear();
    for (int i = 0; i < bucket_count; i++) {
        sort(buckets[i].begin(), buckets[i].end());
        for (int j = 0; j < buckets[i].size(); j++) {
            arr.push_back(buckets[i][j]);
            intercambios++;
        }
    }
}   

int main() {
    vector<int> datos = {64, 25, 12, 22, 11, 90, 45, 33};
    vector<int> datos_original = datos;
    int intercambios = 0;
    int comparaciones = 0;
    auto start_time = chrono::high_resolution_clock::now();
    bubble_sort(datos, intercambios, comparaciones);
    datos = datos_original; // Reset the array before the next sort
    auto end_time = chrono::high_resolution_clock::now();
    cout << "Bubble Sort intercambios: " << intercambios << endl;
    cout << "Bubble Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;

    start_time = chrono::high_resolution_clock::now();
    intercambios = 0;
    comparaciones = 0;
    selection_sort(datos, intercambios, comparaciones);
    datos = datos_original; // Reset the array before the next sort
    end_time = chrono::high_resolution_clock::now();
    cout << "Selection Sort intercambios: " << intercambios << endl;
    cout << "Selection Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;

    start_time = chrono::high_resolution_clock::now();
    intercambios = 0;
    comparaciones = 0;
    insertion_sort(datos, intercambios, comparaciones);
    datos = datos_original; // Reset the array after insertion sort
    end_time = chrono::high_resolution_clock::now();
    cout << "Insertion Sort intercambios: " << intercambios << endl;
    cout << "Insertion Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;

    start_time = chrono::high_resolution_clock::now();
    intercambios = 0;
    comparaciones = 0;
    merge_sort(datos, 0, datos.size() - 1, intercambios, comparaciones);
    datos = datos_original; // Reset the array before the next sort
    end_time = chrono::high_resolution_clock::now();
    cout << "Merge Sort intercambios: " << intercambios << endl;
    cout << "Merge Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;

    start_time = chrono::high_resolution_clock::now();
    intercambios = 0;
    comparaciones = 0;
    quick_sort(datos, 0, datos.size() - 1, intercambios, comparaciones);
    datos = datos_original; // Reset the array before the next sort
    end_time = chrono::high_resolution_clock::now();
    cout << "Quick Sort intercambios: " << intercambios << endl;
    cout << "Quick Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;

    start_time = chrono::high_resolution_clock::now();
    intercambios = 0;
    comparaciones = 0;
    heap_sort(datos, intercambios, comparaciones);
    datos = datos_original; // Reset the array before the next sort
    end_time = chrono::high_resolution_clock::now();
    cout << "Heap Sort intercambios: " << intercambios << endl;
    cout << "Heap Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;

    start_time = chrono::high_resolution_clock::now();
    intercambios = 0;
    comparaciones = 0;
    bucket_sort(datos, intercambios, comparaciones);
    datos = datos_original; // Reset the array before the next sort
    end_time = chrono::high_resolution_clock::now();
    cout << "Bucket Sort intercambios: " << intercambios << endl;
    cout << "Bucket Sort comparaciones: " << comparaciones << endl;
    cout << "Tiempo transcurrido: " << chrono::duration<double>(end_time - start_time).count() << " segundos" << endl;
    return 0;
}